### ISSUE TITLE
Notifications: if the app is stopped, no notifications are displayed

### ISSUE THREAD

**Issue by mzorz** [wordpress-mobile/WordPress-Android#4657]
### Expected behavior
I should be receiving notifications from WP if I have the app installed, no matter what the running state for the app itself is.

### Actual behavior
- after stopping the app, no notifications are received / processed / shown to the user. They start appearing after I first launch the app again.
- if you power off your device and turn it back on, but never launch WordPress manually, you never get any notifs.

### Steps to reproduce the behavior
CASE A) (notifications are received)
1. Log in to the Android WordPress app
2. Have someone like/comment on a blog post of yours
3. See the notification arrive at your device

CASE B) (notifications are not received)
1. Turn your phone off and then back on, but do not launch the WordPress App. Alternatively, you can also go to Apps-> WordPress and then hit `FORCE STOP`.
2. Have someone like/comment on a blog post of yours
3. See the notification NEVER arrive at your device.

**Comment by daniloercoli** [comment]
> CASE B) (notifications are not received)
> 1. Turn your phone off and then back on, but do not launch the WordPress App.

This sound weird...I'm usually turning the device off at night, and get notifications in the morning, as soon as I turn it back on. To be honest, I experienced this issue with other apps too (our app, Whatzapp, etc) and found out that you have to wait bit before receiving the PNs.

**Comment by daniloercoli** [comment]
> Alternatively, you can also go to Apps-> WordPress and then hit FORCE STOP.

It's normal that you stop receiving PNs if you FORCE STOP the app. 
You have to implement an auto-restarting service, hook on the device timer capability and restart a service at predefined interval, or you have to wait that the Android System restart the app itself.

Also tested with FB Messenger. If you force stop it, you will not receive messages. Wait a bit, until their service is restarted, and you will receive all the PNs. If you go into Apps->Messenger you will see the uptime...

**Comment by mzorz** [comment]
All good, I knew how it works as had other apps do the same in the past. We should have the service implemented so PNs start getting in, which we don't have at the moment, so I decided to post this issue. The phone off case for me doesn't ever work, so it's the same situation at least in my case. For both cases, registering for on boot event and a service will work. Decided to write the issue to work on it in the week as time allows :+1:

**Comment by daniloercoli** [comment]
>The phone off case for me doesn't ever work, so it's the same situation at least in my case. For both >cases, registering for on boot event and a service will work

Right. Not 100% sure of this,  but I don't think we should write a service to fix the "phone off" case. The device should handle this case itself. 

For the other case (FORCE STOP) are we really sure we want to implement it? I found annoying those apps that auto restart after a bit, if I had turned them off. FB/WH does this, and they hurt my feelings :)

**Comment by mzorz** [comment]
Fair enough about restarting on force stop which as you say might be an undesired behavior. I would still start it on boot up as if I had the app running I would still want the notifications to come in asap, just like Gmail or WhatsApp do.
I'm planning on building a service to handle like/approve notif quick actions without having to make the app show up anyway, after we have this it will be easier to just schedule,/register this service to start on boot

**Comment by daniloercoli** [comment]
> Fair enough about restarting on force stop which as you say might be an undesired behavior.

That was my 2 cents. Unfortunately, both WH and FB (and maybe others)  seem to restart their services after a bit. Maybe it's a common practice now.  Don't have strong opinion on that.

> I would still start it on boot up as if I had the app running I would still want the notifications to come in asap, just like Gmail or WhatsApp do.

I had this problem with WH till a couple of weeks ago. Don't know if they changed something or it was fixed at device level, but now I receive notifications as soon as I start the device in the morning.
Before the fix I had to start the app manually every morning (annoying).
The thing that let me think is that I don't have this problem with our app. I receive notifications as soon as I start the device. Will try tomorrow.

> I'm planning on building a service to handle like/approve notif quick actions without having to make the app show up anyway, after we have this it will be easier to just schedule,/register this service to start on boot

Cool!

**Comment by mzorz** [comment]
I appreciate your comments a lot! Will try and see how it works this week :+1: will definitely ping you for review too :)

**Comment by daniloercoli** [comment]
Case B)
It could be a side of Doze mode https://developer.android.com/training/monitoring-device-state/doze-standby.html#using_gcm

We should try whitelisting our app and see if does solve the problem before triggering the correct event programatically.

**Comment by mzorz** [comment]
Ok having done some more research I found out:

- as per my tests, wasn't able to reproduce phone off/phone on scenario consistently. I did receive PNs after turning the phone back on (maybe had to wait a bit more, but eventually PNs would be delivered), and so have others reported.
- the _force stop_ scenario works like this by design, so it is expected that the app won't receive any new PNs while stopped.

Closing this as this is actually a non-issue.

More info here:
https://groups.google.com/forum/#!msg/android-gcm/MqVlA3Sj26c/6O4TdU1pm0UJ
